File: Processing_M-Files/spike3FindPeakOffsets.m
% peakList = list of freqs found using curve fit from 3 points inc maxFreq

%#eml

function peakList = spike3FindPeakOffsets(fftSize,spectra,maxFreq)

    % Use curve fit on maximum frequencies, 
    % 1 point on either side to find peak
    % Find up to first 50 frequencies (should be enough)
    peakList = [];
    peakList1=zeros(50); %Initialize array to hold found frequencies
    tempPosition=0; % position in lists of last frequency

    for i=2:fftSize, %Cycle through frequency bins in setFreqMax1 
        %(ignore start to consider two rows. Never a frequency in bin 1=0Hz)
        if(maxFreq(i)==1), % if a peak was identified...
            % increment peak counter
            tempPosition = tempPosition + 1;
            % find peak position in units of bin spacing
            peakList1(tempPosition) = i 
                + peakSpikeOffset(spectra(i-1),spectra(i),spectra(i+1));
        end
    end
    if (tempPosition ~= 0),
        % remove trailing zeros in freqList, freqConsistency
        peakList = peakList1(1:tempPosition);
    end
end

File: Embedded_C_for_Spectrum_Digital_eZdsp_28335/spike3FindPeakOffsets.c
/*
 * spike3FindPeakOffsets.c
 *
 *  Created on: Apr 11, 2010
 *      Author: Jeff
 */
#include "CableMonitoring.h"

// peakList = list of freqs found using spike fit from 3 points inc freqMax

float peakSpikeOffset(int x1,int x2,int x3);

void spike3FindPeakOffsets(){
    // Use curve fit on maximum frequencies, 1 point on either side to find peak
    // Find up to first 50 frequencies (should be enough)
    
    int tempPosition=0; // position in lists of last frequency
    int i;

    for (i=1;i<SPECTRA_LENGTH;i++){ //Cycle through frequency bins in setFreqMax1 (ignore start to consider two rows. Never a frequency in bin 1=0Hz)
        if(freqMax[i]==1){ // if a peak was identified...
            // increment peak counter
            tempPosition = tempPosition + 1;
            if (tempPosition == 50) return; // Maximum number of peaks already reached
            // find peak position in units of bin spacing. Use i + 1 for proper bin ID vs. array indexes
            peakList[tempPosition] = ((float) (i+1)) + peakSpikeOffset(spectraTable[i-1],spectraTable[i],spectraTable[i+1]);
        }
    }
}

